Hw6 Redis caching and metrics monitoring

VfourTwenty opened the pull request.

VfourTwenty committed:
added redis caching and monitoring with prometheus

StasNikonov commented on src/utils/redisClient.js:
гарний базовий redisClient

StasNikonov commented on src/services/weatherService.js:
Чудова інтеграція кешування у сервіс

StasNikonov commented on src/services/weatherService.js:
```diff
@@ -0,0 +1,103 @@
+    static async fetchWeather(city, provider = null) {
+        const cacheKey = `weather:${city.toLowerCase()}`;
+        // 1. Try Redis cache
+        const cached = await redisClient.get(cacheKey);
```
Можна спробувати обгорнути Redis у try/catch, аби не ламати всю відповідь, якщо Redis тимчасово недоступний

StasNikonov commented on src/app.js:
```diff
@@ -39,26 +44,6 @@ app.use(function(err, req, res, next) {
 });
 
-
-cron.schedule('0 * * * *', async () => {
```
Чудове рішення винести логіку Cron у окремий файл

StasNikonov reviewed (approved):
Гарна робота з інтеграцією кешування у проєкт. Як варіант покращення - можна додати Prometheus і Grafana для красивої візуалізації метрик. Також варто додатково покрити тестами логіку кешування у сервісі

VfourTwenty commented:
> Гарна робота з інтеграцією кешування у проєкт. Як варіант покращення - можна додати Prometheus і Grafana для красивої візуалізації метрик. Також варто додатково покрити тестами логіку кешування у сервісі

метрики доступні на https://localhost:3001/metrics (or other baseUrl/metrics depending on env), тести по кешуванню додам після апруву основних тестів

darkravchuk reviewed (commented):
Nice implementation of caching logic!
I would just add some improvements suggestions.

darkravchuk commented on src/services/weatherService.js:
```diff
@@ -0,0 +1,121 @@
+     * @returns {Promise<any>}
+     */
+
+    static logPath = require('path').join(__dirname, '../../logs/weatherProvider.log');
```
Move configuration logPath, loggingEnabled to a config file or environment variables

VfourTwenty replied:
will go with env variables when logging is properly refactored in task 5

darkravchuk commented on src/services/weatherService.js:
```diff
@@ -0,0 +1,121 @@
+
+    // possible to pass a specific provider, or rely on the chain of responsibility
+    static async fetchWeather(city, provider = null) {
+        const cacheKey = `weather:${city.toLowerCase()}`;
```
The cache key (weather:${city.toLowerCase()}) may be inconsistent with special characters or spaces in city names.
Suggestion: Normalize the city name trim spaces and replace special characters for consistent cache keys

darkravchuk commented on src/utils/redisClient.js:
```diff
@@ -0,0 +1,11 @@
+const redis = require('redis');
+
+const client = redis.createClient(); // defaults to localhost:6379
```
The Redis client uses default settings (localhost:6379), which may not be secure in production.
Use environment variable for Redis URL

darkravchuk commented on src/services/weatherService.js:
```diff
@@ -0,0 +1,121 @@
+// need to abstract from sequelize operators
+const {Op} = require("sequelize");
+const { logProviderResponse } = require('../utils/logger');
+const redisClient = require('../utils/redisClient');
```
Hard-coded dependencies redisClient make unit testing difficult.
Use dependency injection via constructor to allow mocking in tests

VfourTwenty replied:
good idea, will do once my usage of dependency injection gets approved

darkravchuk commented on src/utils/redisClient.js:
```diff
@@ -0,0 +1,11 @@
+
+client.on('error', (err) => console.error('Redis Client Error', err));
+
+(async () => {
```
Using IIFE to connect, could lead to race conditions if WeatherService uses it before the connection is established. Maybe expose a connect method and ensure WeatherService awaits it before use

VfourTwenty committed:
moved interfaces to domain layer

VfourTwenty committed:
updated weatherService to use redis and prometheus

VfourTwenty committed:
added docker services for redis and prometherus and necessary config

VfourTwenty committed:
updated integration tests to use redis cache

VfourTwenty committed:
updated backend service dependency

VfourTwenty committed:
added redis to e2e tests

VfourTwenty committed:
now reading readis TTL from env

VfourTwenty committed:
added unit tests for weather service to check interations with cache and metrics

VfourTwenty committed:
fixed linter issues

VfourTwenty committed:
removed the old metrics setup

VfourTwenty committed:
added REDIS_PORT env var to CI

VfourTwenty committed:
added a missing redis dependency for integration tests in docker

VfourTwenty committed:
introduced more advanced string normalization for consistency in weather cache keys

VfourTwenty committed:
added string normalization unit tests

VfourTwenty committed:
renamed client to redisClient for clarity

VfourTwenty committed:
refactored redis connection strategy to prevent race conditions

VfourTwenty commented:
Changes summary:
1) Introduced service interfaces in the domain layer so that weatherService with caching and metrics follows a base contract + allows for easier extension of other services
2) Redis client and metrics counters are now injectable into the weather service 
3) Redis is now configurable via env variables
4) Added more advanced string normalization to account for cache key consistency
5) Refactored Redis connection strategy so that the connection is ensured to be established before the app starts to prevent race conditions
6) Added new unit tests for weather service and string normalization
7) Configured all necessary services in docker

VfourTwenty committed:
fixed a comment

VfourTwenty committed:
provided abstractions for cache and metrics

VfourTwenty committed:
updated dependencies

VfourTwenty committed:
updated tests and mocks

VfourTwenty committed:
removed deprecated prom and redis clients from utils

VfourTwenty committed:
fixed a linting issue

vasyl-khrapko-liven commented on src/common/cache/redis/redisClient.js:
```diff
@@ -0,0 +1,37 @@
+const config = require('../../../config/index');
+const retry = require('../../../utils/retry');
+
+// Create the client ONCE
```
Чому залишила такий коментар? Як на мене такі коментарі зайві

vasyl-khrapko-liven commented on src/domain/interfaces/services/emailServiceInterface.js:
```diff
@@ -0,0 +1,15 @@
+class IEmailService
```
Визначати інтерфейси для сервісів, які є частиною це вже оверхед, у тебе сильно збільшується кількість файлів, які потрібно редагувати + додавати при створенні нових сутностей + треба тримати той рівень абстракції для деталей, які в тебе нестабільні, який дозволить продовжувати підтримувати розробку і не робити абстрактним весь проєкт

У сервісах ти визначаєш бізнес сценарії, навряд у тебе буде декілька варіантів відправки листів або роботи з підписками, а вже деталі інструментів роботи над ними нестабільні і можуть змінюватись 

У Роберта Мартіна в "Чистій архітектурі" є про це гарна діаграмка
<img width="800" height="800" alt="image" src="https://github.com/user-attachments/assets/6e7c87fd-c159-4e93-8f0d-89257ab9496f" />

vasyl-khrapko-liven commented on src/services/weatherService.js:
```diff
@@ -1,23 +1,122 @@
-// const { logProviderResponse } = require('../utils/logger');
+const IWeatherService = require('../domain/interfaces/services/weatherServiceInterface');
+// const config = require('../config/index');
+const { normalizeString } = require('../utils/strings');
```
Коментарі дуже заважають переглядати код на рев'ю, не дуже окей практика залишати "чорнові" або попередні варіанти, вони будуть збивати колег при взаємодії з твоїм кодом і його підтримці

vasyl-khrapko-liven commented on src/services/weatherService.js:
```diff
@@ -1,23 +1,122 @@
-    async fetchWeather(city) 
+    async fetchWeather(city)
     {
+        if (!city)
```
відступів між рядками немає, це читається складно(

vasyl-khrapko-liven commented on src/common/metrics/metricsProviderInterface.js:
```diff
@@ -0,0 +1,53 @@
+class IMetricsProvider
```
Спроба зробити абстракцію над провайдером метрик хороша, нічого змінювати не потрібно

Єдине що хочу підкреслити, що в реальному проєкті часто сервіси для метрик можуть бути специфічні й використовувати різні підходи і підтримка абстракції може бути складною задачею, яка того не варта

vasyl-khrapko-liven commented on src/services/weatherService.js:
```diff
@@ -1,23 +1,122 @@
+        this.cacheProvider = cacheProvider;
+        this.metricsProvider = metricsProvider;
+
+        this.metricsProvider.registerMetric('counter', 'weatherCacheHits','Total number of weather cache hits');
```
тут сетап метрик виконує клас WeatherServiceWithCacheAndMetrics - це не його відповідальність. Було б добре заздалегідь засетапити метрики в metrics модулі, а тут лише використовувати metrics сервіс для інкрементації значень

vasyl-khrapko-liven commented on src/services/weatherService.js:
```diff
@@ -1,23 +1,122 @@
+        }
+        if (cachedWeather)
+        {
+            this.metricsProvider.incrementCounter('weatherCacheHits');
```
константні ключі було б добре тримати в колекції ключей метрик

vasyl-khrapko-liven commented on src/services/weatherService.js:
```diff
@@ -1,23 +1,122 @@
+// }
 
-class WeatherService 
+class WeatherServiceWithCacheAndMetrics extends IWeatherService
```
Неймінг дуже розмиває відповідальнісь цього класу, його задача робота з погодою, а сайд ефекти (логування, запис метрик, робота з кешем) - це деталі реалізації WeatherService, тому краще було б залишити старий неймінг

vasyl-khrapko-liven commented on test/unit/services/weatherService.test.js:
```diff
@@ -1,24 +1,56 @@
+        metricsProviderMock.reset();
+    });
+
+    it('should return true for success and weather data for a valid city saving the data to cache and incrementing cache miss counter', async () =>
```
Робота з метриками - це деталі реалізації, так само як і логування - варто тестувати бізнес логіку, а не деталі реалізації в тестах

Щодо кешей, якщо тобі важлива робота з кешем (наприклад, тобі важлива швидкість виконання методу), це окей тестувати сценарії роботи з кешем

vasyl-khrapko-liven reviewed (approved):
Робота виконана добре, додано роботу з кешем (враховані абстракції для роботи з кешем), також додано моніторинг операцій, пов'язаних з кешем через Prometheus

Є мінорні зауваження по коду - можна виправляти і переходити до наступного дз

VfourTwenty committed:
added a comment for external provider

VfourTwenty committed:
added logging for email providers chain

VfourTwenty committed:
updated email provider manager unit tests

VfourTwenty committed:
Merge branch 'main' into hw6-databases

VfourTwenty committed:
removed a comment

VfourTwenty committed:
renamed weather service and removed some comments

VfourTwenty committed:
separated matrics setup from weather service and refactored unit tests

VfourTwenty committed:
introduced constants for metrics keys
